File: bpack.js
function bpack(size)
{
	this.buf = new Buffer(size);
	this.offset = 0;
}

bpack.prototype.write_int32 = function(val)
{
	this.buf.writeInt32LE(val, this.offset);
	this.offset += 4;
};

bpack.prototype.read_int32 = function(buffer, offset)
{
	return buffer.readInt32LE(offset);
};

bpack.prototype.write_uint32 = function(val)
{
	this.buf.writeUInt32LE(val, this.offset);
	this.offset += 4;
};

bpack.prototype.read_uint32 = function(buffer, offset)
{
	return buffer.readUInt32LE(offset);
};

bpack.prototype.write_int16 = function(val)
{
	this.buf.writeInt16LE(val, this.offset);
	this.offset += 2;
};

bpack.prototype.read_int16 = function(buffer, offset)
{
	return buffer.readInt16LE(offset);
};

bpack.prototype.write_uint16 = function(val)
{
	this.buf.writeUInt16LE(val, this.offset);
	this.offset += 2;
};

bpack.prototype.read_uint16 = function(buffer, offset)
{
	return buffer.readUInt16LE(offset);
};


bpack.prototype.write_uint8 = function(val)
{
	this.buf.writeUInt8(val, this.offset);
	this.offset += 1;
};

bpack.prototype.read_uint8 = function(buffer, offset)
{
	return buffer.readUInt8(offset);
};

bpack.prototype.write_float = function(val)
{
	this.buf.writeFloatLE(val, this.offset);
	this.offset += 4;
};

bpack.prototype.read_float = function(buffer, offset)
{
	return buffer.readFloatLE(offset);
};

bpack.prototype.write_double = function(val)
{
	this.buf.writeDoubleLE(val, this.offset);
	this.offset += 8;
};

bpack.prototype.read_double = function(buffer, offset)
{
	return buffer.readDoubleLE(offset);
};

bpack.prototype.write_string = function(val)
{
	var len = Buffer.byteLength(val, 'utf8');
	this.buf.writeInt32LE(len, this.offset);
	this.offset += 4;
	this.buf.write(val, this.offset, 'utf8');
	this.offset += len;
};

bpack.prototype.read_string = function(buffer, offset)
{
	var len = buffer.readInt32LE(offset);
	offset += 4;
	return buffer.toString('utf8', offset, offset + len);
};

bpack.prototype.write_buffer = function(buffer)
{
	this.buf.writeInt32LE(buffer.length, this.offset);
	this.offset += 4;
	buffer.copy(this.buf, this.offset);
	this.offset += buffer.length;
};

bpack.prototype.read_buffer = function(buffer, offset)
{
	var len = buffer.readInt32LE(offset);
	offset += 4;
	return buffer.slice(offset, offset + len);
};

bpack.prototype.write_buffer_array = function(arr)
{
	this.buf.writeInt32LE(arr.length, this.offset);
	this.offset += 4;
	for (var i = 0; i < arr.length; i++)
	{
		this.write_buffer(arr[i]);
	}
};

bpack.prototype.read_buffer_array = function(buffer, offset)
{
	var len = buffer.readInt32LE(offset);
	offset += 4;
	var result = [];
	for (var i = 0; i < len; i++)
	{
		result.push(this.read_buffer(buffer, offset));
		offset += result[i].length + 4;
	}
	return result;
};

bpack.prototype.write_int32_array = function(arr)
{
	this.buf.writeInt32LE(arr.length, this.offset);
	this.offset += 4;
	for (var i = 0; i < arr.length; i++)
	{
		this.write_int32(arr[i]);
	}
};

bpack.prototype.read_int32_array = function(buffer, offset)
{
	var len = buffer.readInt32LE(offset);
	offset += 4;
	var result = [];
	for (var i = 0; i < len; i++)
	{
		result.push(this.read_int32(buffer, offset));
		offset += 4;
	}
	return result;
};

bpack.prototype.write_uint32_array = function(arr)
{
	this.buf.writeInt32LE(arr.length, this.offset);
	this.offset += 4;
	for (var i = 0; i < arr.length; i++)
	{
		this.write_uint32(arr[i]);
	}
};

bpack.prototype.read_uint32_array = function(buffer, offset)
{
	var len = buffer.readInt32LE(offset);
	offset += 4;
	var result = [];
	for (var i = 0; i < len; i++)
	{
		result.push(this.read_uint32(buffer, offset));
		offset += 4;
	}
	return result;
};

bpack.prototype.write_int16_array = function(arr)
{
	this.buf.writeInt32LE(arr.length, this.offset);
	this.offset += 4;
	for (var i = 0; i < arr.length; i++)
	{
		this.write_int16(arr[i]);
	}
};

bpack.prototype.read_int16_array = function(buffer, offset)
{
	var len = buffer.readInt32LE(offset);
	offset += 4;
	var result = [];
	for (var i = 0; i < len; i++)
	{
		result.push(this.read_int16(buffer, offset));
		offset += 2;
	}
	return result;
};

bpack.prototype.write_uint16_array = function(arr)
{
	this.buf.writeInt32LE(arr.length, this.offset);
	this.offset += 4;
	for (var i = 0; i < arr.length; i++)
	{
		this.write_uint16(arr[i]);
	}
};

bpack.prototype.read_uint16_array = function(buffer, offset)
{
	var len = buffer.readInt32LE(offset);
	offset += 4;
	var result = [];
	for (var i = 0; i < len; i++)
	{
		result.push(this.read_uint16(buffer, offset));
		offset += 2;
	}
	return result;
};

bpack.prototype.write_uint8_array = function(arr)
{
	this.buf.writeInt32LE(arr.length, this.offset);
	this.offset += 4;
	for (var i = 0; i < arr.length; i++)
	{
		this.write_uint8(arr[i]);
	}
};

bpack.prototype.read_uint8_array = function(buffer, offset)
{
	var len = buffer.readInt32LE(offset);
	offset += 4;
	var result = [];
	for (var i = 0; i < len; i++)
	{
		result.push(this.read_uint8(buffer, offset));
		offset += 1;
	}
	return result;
};

bpack.prototype.write_float_array = function(arr)
{
	this.buf.writeInt32LE(arr.length, this.offset);
	this.offset += 4;
	for (var i = 0; i < arr.length; i++)
	{
		this.write_float(arr[i]);
	}
};

bpack.prototype.read_float_array = function(buffer, offset)
{
	var len = buffer.readInt32LE(offset);
	offset += 4;
	var result = [];
	for (var i = 0; i < len; i++)
	{
		result.push(this.read_float(buffer, offset));
		offset += 4;
	}
	return result;
};

bpack.prototype.write_double_array = function(arr)
{
	this.buf.writeInt32LE(arr.length, this.offset);
	this.offset += 4;
	for (var i = 0; i < arr.length; i++)
	{
		this.write_double(arr[i]);
	}
};

bpack.prototype.read_float_array = function(buffer, offset)
{
	var len = buffer.readInt32LE(offset);
	offset += 4;
	var result = [];
	for (var i = 0; i < len; i++)
	{
		result.push(this.read_double(buffer, offset));
		offset += 8;
	}
	return result;
};

bpack.prototype.write_string_array = function(arr)
{
	this.buf.writeInt32LE(arr.length, this.offset);
	this.offset += 4;
	for (var i = 0; i < arr.length; i++)
	{
		this.write_string(arr[i]);
	}
};

bpack.prototype.read_string_array = function(buffer, offset)
{
	var len = buffer.readInt32LE(offset);
	offset += 4;
	var result = [];
	for (var i = 0; i < len; i++)
	{
		var str = this.read_string(buffer, offset);
		result.push(str);
		offset += 4 + Buffer.byteLength(str, 'utf8');
	}
	return result;
};



module.exports = bpack;

File: BPack.cs
using System;
using System.IO;
using System.Text;

public class BPack : IDisposable
{
    private MemoryStream buf;
    private BinaryWriter writer;
    private BinaryReader reader;
    private int readOffset;
    private int writeOffset;

    public BPack(int size)
    {
        buf = new MemoryStream(size);
        writer = new BinaryWriter(buf);
        reader = new BinaryReader(buf);
    }

    public BPack(byte[] buffer)
    {
        buf = new MemoryStream(buffer);
        writer = new BinaryWriter(buf);
        reader = new BinaryReader(buf);
    }

    // ===== WRITE METHODS =====
    public void WriteInt32(int val)
    {
        writer.Write(val);
        writeOffset += 4;
    }

    public void WriteUInt32(uint val)
    {
        writer.Write(val);
        writeOffset += 4;
    }

    public void WriteInt16(short val)
    {
        writer.Write(val);
        writeOffset += 2;
    }

    public void WriteUInt16(ushort val)
    {
        writer.Write(val);
        writeOffset += 2;
    }

    public void WriteByte(byte val)
    {
        writer.Write(val);
        writeOffset += 1;
    }

    public void WriteFloat(float val)
    {
        writer.Write(val);
        writeOffset += 4;
    }

    public void WriteDouble(double val)
    {
        writer.Write(val);
        writeOffset += 8;
    }

    public void WriteString(string val)
    {
        byte[] bytes = Encoding.UTF8.GetBytes(val);
        WriteInt32(bytes.Length);
        writer.Write(bytes);
        writeOffset += bytes.Length;
    }

    public void WriteBuffer(byte[] buffer)
    {
        WriteInt32(buffer.Length);
        writer.Write(buffer);
        writeOffset += buffer.Length;
    }

    // ===== WRITE ARRAY METHODS =====
    public void WriteInt32Array(int[] arr)
    {
        WriteInt32(arr.Length);
        foreach (var val in arr)
            WriteInt32(val);
    }

    public void WriteUInt32Array(uint[] arr)
    {
        WriteInt32(arr.Length);
        foreach (var val in arr)
            WriteUInt32(val);
    }

    public void WriteInt16Array(short[] arr)
    {
        WriteInt32(arr.Length);
        foreach (var val in arr)
            WriteInt16(val);
    }

    public void WriteUInt16Array(ushort[] arr)
    {
        WriteInt32(arr.Length);
        foreach (var val in arr)
            WriteUInt16(val);
    }

    public void WriteByteArray(byte[] arr)
    {
        WriteInt32(arr.Length);
        writer.Write(arr);
        writeOffset += arr.Length;
    }

    public void WriteFloatArray(float[] arr)
    {
        WriteInt32(arr.Length);
        foreach (var val in arr)
            WriteFloat(val);
    }

    public void WriteDoubleArray(double[] arr)
    {
        WriteInt32(arr.Length);
        foreach (var val in arr)
            WriteDouble(val);
    }

    public void WriteStringArray(string[] arr)
    {
        WriteInt32(arr.Length);
        foreach (var val in arr)
            WriteString(val);
    }

    public void WriteBufferArray(byte[][] arr)
    {
        WriteInt32(arr.Length);
        foreach (var val in arr)
            WriteBuffer(val);
    }

    // ===== READ METHODS =====
    public int ReadInt32()
    {
        int val = reader.ReadInt32();
        readOffset += 4;
        return val;
    }

    public uint ReadUInt32()
    {
        uint val = reader.ReadUInt32();
        readOffset += 4;
        return val;
    }

    public short ReadInt16()
    {
        short val = reader.ReadInt16();
        readOffset += 2;
        return val;
    }

    public ushort ReadUInt16()
    {
        ushort val = reader.ReadUInt16();
        readOffset += 2;
        return val;
    }

    public byte ReadByte()
    {
        byte val = reader.ReadByte();
        readOffset += 1;
        return val;
    }

    public float ReadFloat()
    {
        float val = reader.ReadSingle();
        readOffset += 4;
        return val;
    }

    public double ReadDouble()
    {
        double val = reader.ReadDouble();
        readOffset += 8;
        return val;
    }

    public string ReadString()
    {
        int length = ReadInt32();
        byte[] bytes = reader.ReadBytes(length);
        readOffset += length;
        return Encoding.UTF8.GetString(bytes);
    }

    public byte[] ReadBuffer()
    {
        int length = ReadInt32();
        byte[] buffer = reader.ReadBytes(length);
        readOffset += length;
        return buffer;
    }

    // ===== READ ARRAY METHODS =====
    public int[] ReadInt32Array()
    {
        int length = ReadInt32();
        int[] arr = new int[length];
        for (int i = 0; i < length; i++)
            arr[i] = ReadInt32();
        return arr;
    }

    public uint[] ReadUInt32Array()
    {
        int length = ReadInt32();
        uint[] arr = new uint[length];
        for (int i = 0; i < length; i++)
            arr[i] = ReadUInt32();
        return arr;
    }

    public short[] ReadInt16Array()
    {
        int length = ReadInt32();
        short[] arr = new short[length];
        for (int i = 0; i < length; i++)
            arr[i] = ReadInt16();
        return arr;
    }

    public ushort[] ReadUInt16Array()
    {
        int length = ReadInt32();
        ushort[] arr = new ushort[length];
        for (int i = 0; i < length; i++)
            arr[i] = ReadUInt16();
        return arr;
    }

    public byte[] ReadByteArray()
    {
        int length = ReadInt32();
        byte[] arr = reader.ReadBytes(length);
        readOffset += length;
        return arr;
    }

    public float[] ReadFloatArray()
    {
        int length = ReadInt32();
        float[] arr = new float[length];
        for (int i = 0; i < length; i++)
            arr[i] = ReadFloat();
        return arr;
    }

    public double[] ReadDoubleArray()
    {
        int length = ReadInt32();
        double[] arr = new double[length];
        for (int i = 0; i < length; i++)
            arr[i] = ReadDouble();
        return arr;
    }

    public string[] ReadStringArray()
    {
        int length = ReadInt32();
        string[] arr = new string[length];
        for (int i = 0; i < length; i++)
            arr[i] = ReadString();
        return arr;
    }

    public byte[][] ReadBufferArray()
    {
        int length = ReadInt32();
        byte[][] arr = new byte[length][];
        for (int i = 0; i < length; i++)
            arr[i] = ReadBuffer();
        return arr;
    }

    // ===== UTILITY METHODS =====
    public byte[] GetBuffer()
    {
        writer.Flush();
        return buf.ToArray();
    }

    public void ResetRead()
    {
        buf.Seek(0, SeekOrigin.Begin);
        readOffset = 0;
    }

    public void Dispose()
    {
        writer?.Dispose();
        reader?.Dispose();
        buf?.Dispose();
    }
}
